dm/master/etcd.go

```diff
@@ -59,3 +70,131 @@ func startEtcd(masterCfg *Config,
+// learn from https://github.com/pingcap/pd/blob/37efcb05f397f26c70cda8dd44acaa3061c92159/server/join/join.go#L44.
+//
+// when setting `initial-cluster` explicitly to bootstrap a new cluster:
+// - if local persistent data exist, just restart the previous cluster (in fact, it's not bootstrapping).
+// - if local persistent data not exist, just bootstrap the cluster as a new cluster.
+//
+// when setting `join` to join an existing cluster (without `initial-cluster` set):
+// - if local persistent data exists (in fact, it's not join):
+//   - just restart if `member` already exists (already joined before)
+//   - read `initial-cluster` back from local persistent data to restart (just like bootstrapping)
+// - if local persistent data not exist:
+//   1. fetch member list from the cluster to check if we can join now.
+//   2. call `member add` to add the member info into the cluster.
+//   3. generate config for join (`initial-cluster` and `initial-cluster-state`).
+//   4. save `initial-cluster` in local persistent data for later restarting.
+//
+// NOTE: A member can't join to another cluster after it has joined a previous one.
+func prepareJoinEtcd(cfg *Config) error {
+	// no need to join
+	if cfg.Join == "" {
+		return nil
+	}
+
+	// try to join self, invalid
+	if cfg.Join == cfg.AdvertisePeerUrls {
+		return terror.ErrMasterJoinEmbedEtcdFail.Generate(fmt.Sprintf("join self %s is forbidden", cfg.Join))
+	}
+
+	// restart with previous data, no `InitialCluster` need to set
+	if isDataExist(filepath.Join(cfg.DataDir, "member")) {
+		cfg.InitialCluster = ""
+		cfg.InitialClusterState = embed.ClusterStateFlagExisting
+		return nil
+	}
+
+	// join with persistent data
+	joinFP := filepath.Join(cfg.DataDir, "join")
+	if s, err := ioutil.ReadFile(joinFP); err != nil {
+		if !os.IsNotExist(err) {
+			return terror.ErrMasterJoinEmbedEtcdFail.Delegate(err, "read persistent join data")
+		}
+	} else {
+		cfg.InitialCluster = strings.TrimSpace(string(s))
+		cfg.InitialClusterState = embed.ClusterStateFlagExisting
+		log.L().Info("using persistent join data", zap.String("file", joinFP), zap.String("data", cfg.InitialCluster))
+		return nil
+	}
+
+	// if without previous data, we need a client to contact with the existing cluster.
+	client, err := clientv3.New(clientv3.Config{
+		Endpoints:   strings.Split(cfg.Join, ","),
+		DialTimeout: etcdutil.DefaultDialTimeout,
+	})
+	if err != nil {
+		return terror.ErrMasterJoinEmbedEtcdFail.Delegate(err, fmt.Sprintf("create etcd client for %s", cfg.Join))
+	}
+	defer client.Close()
+
+	// `member list`
+	listResp, err := etcdutil.ListMembers(client)
+	if err != nil {
+		return terror.ErrMasterJoinEmbedEtcdFail.Delegate(err, fmt.Sprintf("list member for %s", cfg.Join))
+	}
+
+	// check members
+	for _, m := range listResp.Members {
+		if m.Name == "" {
+			return terror.ErrMasterJoinEmbedEtcdFail.Generate("there is a member that has not joined successfully")
```

Comment:
this means other member joined failed, but why prevent this member join? And how to fix this error?

Reply:
add some comments to explain it, and update the error message in https://github.com/pingcap/dm/pull/350/commits/b0f53328e849e95757c15fe7f599e7115b04258e.

later maybe we can embed etcd client in dmctl?